archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,275 @@
+          norm_cast,
+          exact ceil_le_floor_add_one _,
+          apply le_trans _ t,
+          exact le_ceil (sqrt (1 + ↑n) - 1),
+          suffices : sqrt (4 + 2 * n) ≤ 2 * (sqrt (1 + n) - 3),
+            linarith,
+            exact radical_inequality hn },
+    { rw hx,
+      exact floor_le _,},
+    { rw [floor_nonneg, le_sub, sub_zero, le_sqrt],
+      all_goals {norm_cast, linarith}},
+end
+
+lemma exists_triplet_summing_to_squares (n : ℕ) (hn : n ≥ 100): (∃ (a b c : ℕ),
+n ≤ a ∧ a < b ∧ b < c ∧ c ≤ 2 * n ∧  (∃ (k : ℕ), a + b = k * k) ∧
+(∃ (l : ℕ), c + a = l * l) ∧ (∃ (m : ℕ), b + c = m * m)) :=
+begin
+  --If n ≥ 107, we do not explicitly construct the triplet but use an existence
+  --argument from lemma above.
+  by_cases p : n ≥ 107,
+  have h := exists_numbers_in_interval n p,
+  cases h with l hl,
+  use (2 * l * l - 4 * l),
+  use (2 * l * l + 1),
+  use (2 * l * l + 4 * l),
+  by_cases p : l > 1,
+  refine ⟨_,⟨_,⟨_,⟨_,⟨_, ⟨_, _⟩⟩⟩⟩⟩⟩,
+  { exact nat.le_sub_right_of_add_le hl.1},
+  { exact (2 * l * l).sub_lt_succ (4 * l)},
+  { linarith},
+  { exact hl.2},
+  { use (2 * l - 1),
+    have h₁ : 1 ≤ 2 * l := by linarith,
+    have h₂ : 4 * l ≤ 2 * l * l := by linarith,
+    zify [h₁, h₂],
+    ring_exp_eq },
+  { use (2 * l),
+    have h₁ : 4 * l ≤ 2 * l * l := by linarith,
+    zify [h₁],
+    ring_exp_eq},
+  { use(2 * l + 1),
+    ring_exp_eq,
+  },
+    --Otherwise, if 100 ≤ n < 107, then it suffices to consider explicit
+    --construction of a triplet {a, b, c}, which is constructed by setting l=9
+    --in the argument at the start of the file.
+  { exfalso,
+    simp only [not_lt] at p,
+    interval_cases l,
+    all_goals {linarith},},
+  { refine ⟨126, 163, 198, _, _, _, _, ⟨⟨17, _⟩,⟨⟨18, _⟩,⟨19, _⟩⟩⟩⟩,
+    all_goals {linarith} },
+end
+
+lemma mem_compl_iff_nmem_subset {A : finset ℕ} {a b c n : ℕ}
+(t : A ⊆ finset.range(2 * n + 1) \ finset.range n) (h₁ : n ≤ a)
+(h₂ : a < b) (h₃ : b < c) (h₄ : c ≤ 2 * n) :
+(a ∉ A ↔ a ∈ (finset.range (2 * n + 1) \ finset.range n) \ A) ∧
+(b ∉ A ↔ b ∈ (finset.range (2 * n + 1) \ finset.range n) \ A) ∧
+(c ∉ A ↔ c ∈ (finset.range (2 * n + 1) \ finset.range n) \ A) :=
+begin
+  refine ⟨⟨_,_⟩, ⟨⟨_,_⟩,⟨_,_⟩⟩⟩,
+  { intro ha,
+    simp only [finset.mem_sdiff, finset.mem_range],
+    refine ⟨⟨_,_⟩,ha⟩,
+    repeat {linarith}},
+  { intro ha,
+    simp only [finset.mem_sdiff, finset.mem_range] at ha,
+    exact ha.2 },
+  { intro hb,
+    simp only [finset.mem_sdiff, finset.mem_range],
+    refine ⟨⟨_,_⟩,hb⟩,
+    repeat {linarith}},
+  { intro hb,
+    simp only [finset.mem_sdiff, finset.mem_range] at hb,
+    exact hb.2 },
+   {intro hc,
+    simp only [finset.mem_sdiff, finset.mem_range],
+    refine ⟨⟨_,_⟩,hc⟩,
+    repeat {linarith}},
+  { intro hc,
+    simp only [finset.mem_sdiff, finset.mem_range] at hc,
+    exact hc.2 },
+end
+
+
+theorem IMO_2021_Q1 : ∀ (n : ℕ), n ≥ 100 → ∀ (A ⊆ finset.range(2 * n + 1)\finset.range n),
```

Comment:
Can you use `finset.Ico` here? (Perhaps we even have `finset.Icc`?)

Reply:
I do not think we have ```finset.Icc``` (would it be useful if I defined it and proved some basic lemmas about it?). In the new version, I've used ```finset.Ico```, as you've suggested.

Reply:
I think the concensus is that in the discrete situation having three extra definitions (`Icc`, `Ioc`, `Ioo`) bring with them an API burden whose outweighs the advantages. At least for now.

Reply:
I'm taking care of defining `finset.Icc`, `finset.Ico`, `finset.Ioc` and making them non ℕ-specific in #7987. I think you just shouldn't bother for now and I'll try going through the library to change the uses of `finset.Ico` to the more natural version.